sql: improve statement fingerprinting

xinhaoz opened the pull request:
See individual commits.

Epic: none

Release note: None

xinhaoz committed:
sql: introduce new fmt flag for stmt fingerprinting

This commit introduces the `FmtForFingerprint` bit mask to format
a statement AST into a fingerprint for sql stats. This flag will be
used to more aggressively normalize the query string than
`FmtHideConstants` which was what was previously being used in the
statement normalization for sql statistics tracking.

Epic: none

Release note: None

xinhaoz committed:
builtins: create crdb_internal.statement_fingerprint

TODO msg

xinhaoz committed:
sql/sem/tree: add FmtForFingerprint functions

Add functionality for the `FmtForFingerprint` flag
for AST conversion to fingerprint string.

TODO msg

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119134)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Is this ready for a look? Looks there are some incomplete things (like TODOs are still present, tests missing).

> Reviewed 5 of 5 files at r1, 3 of 3 files at r2, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-:-NqZFY7Q22aUBalHkKrh:bpjlujm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz)

___
*[`-- commits` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-NqZEnwK3Gim0nsRjSAb):*
nit: commit messages should be updated.
___
*[`pkg/sql/sem/tree/format_fingerprint.go` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-NqZEcDx1CCA1wBAIGJP:-NqZEcDx1CCA1wBAIGJQ:b-72lfi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf5db49ede3f4a7acbe3fba8d2c3647b874c90d/pkg/sql/sem/tree/format_fingerprint.go#L9)):*
> ```Go
> // the Business Source License, use of this software will be governed
> // by the Apache License, Version 2.0, included in the file
> // licenses/APL.txt.
> ```

nit: in the second commit adding empty `format_fingerprint.go` file seems suspicious - did you mean to include it into the third commit?


<!-- Sent from Reviewable.io -->

xinhaoz commented:
@yuzefovich ah sorry - this was meant to be a draft. please don't review yet 😅

abarganier committed:
pkg/sql/sem/tree: tests for benchmarking stmt fingerprint cardinality

This commit adds a test mechanism to benchmark stmt fingerprint
cardinality generated from stripping stmt strings of constants.

The test structure enables us to use telemetry data in Splunk as an
input to help measure the before/after impact of changes made to our
algorithm to strip stmt strings of constants.

We intentionally leave out all input data as it may be considered
sensitive. Instead, we provide clear directions & documentation for how
to source data from Splunk and feed it into this test framework.

Finally, the tests are marked as Skip, since we don't want these running
as part of the CI process.

Release note: none

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119134)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz and @yuzefovich)

___
*[`pkg/sql/sem/tree/format_fingerprint.go` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-Nqbt1dr1eQyiluoFUZR:-Nqbt1dr1eQyiluoFUZS:b-9hnk2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/87e59881c0d851d4a126836c6b3f64a4aee17a50/pkg/sql/sem/tree/format_fingerprint.go#L47)):*
> ```Go
> 	ctx.WriteByte(')')
> 	if len(node.Rows) > 1 {
> 		ctx.Printf(", (%s)", arityString(len(node.Rows)-1))
> ```

Are these arity markers worthwhile? Seems like they will cause different fingerprints for similar statements that we'd want to collapse.


<!-- Sent from Reviewable.io -->

xinhaoz committed:
sql/sem/tree: sort update set list

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

abarganier commented on pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go:
```diff
@@ -446,7 +441,7 @@ func (s *stmtStats) mergeStatsLocked(statistics *appstatspb.CollectedStatementSt
 
 	// Setting all metadata fields.
-	if s.mu.data.SensitiveInfo.LastErr == "" && statistics.Key.Failed {
+	if s.mu.data.SensitiveInfo.LastErr == "" && statistics.Stats.SensitiveInfo.LastErr != "" {
```
@xinhaoz What do you think about this updated conditional? If the LastErr is empty, and the stats we're merging has its LastErr populated, we take the LastErr from the stats being merged.

xinhaoz committed:
sql/sem/tree: sorting names in certain cases

todo

xinhaoz committed:
wip

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119134)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @xinhaoz, and @yuzefovich)

___
*[`pkg/sql/sem/tree/format_fingerprint.go` line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-NqnPhBz54VqXBTrSV3B:-NqnPhBz54VqXBTrSV3C:b7bzxmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/74a0b46dc350439d750575e94f511c52a3e22081/pkg/sql/sem/tree/format_fingerprint.go#L45)):*
> ```Go
> 			// for equivalent queries. For example:
> 			// SELECT a, b FROM t
> 			// SELECT b, a FROM t
> ```

These are not equivalent queries. The ordering of the columns makes them different. I'm good with bucketing these queries into the same fingerprint, only pointing out that the comment is not accurate.
___
*[`pkg/sql/sem/tree/format_fingerprint.go` line 148 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/119134#-NqnP2pRBsPE8WEWvDqU:-NqnP2pRBsPE8WEWvDqV:b855vjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/74a0b46dc350439d750575e94f511c52a3e22081/pkg/sql/sem/tree/format_fingerprint.go#L148)):*
> ```Go
> 
> 	v2 := append(make(UpdateExprs, 0, len(*node)), *node...)
> 	sort.Slice(v2, func(i, j int) bool {
> ```

@xinhaoz This is the type of fingerprint formatting that is too expensive to be done on the hot-path of query processing. The expensive part is the memory allocation and the sorting. Note that the memory allocation for the `v2` slice is probably more problematic than the sorting, but neither are good.


<!-- Sent from Reviewable.io -->
